File: Medusa/MedusaCore/Core/Siren/Schema/Type/SirenCustomClass.cpp
// Copyright (c) 2015 fjz13. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.
#include "MedusaCorePreCompiled.h"
#include "SirenCustomClass.h"
#include "Core/Log/Log.h"
#include "Core/Siren/Schema/SirenTextParser.h"
#include "Core/Siren/Schema/SirenAssembly.h"
#include "SirenCustomEnum.h"
#include "Core/IO/Stream/IStream.h"

MEDUSA_BEGIN;

SirenCustomClass::SirenCustomClass()
{

}

SirenCustomClass::~SirenCustomClass(void)
{
	SAFE_DELETE_DICTIONARY_VALUE(mTypes);
	SAFE_DELETE_DICTIONARY_VALUE(mFields);

}

bool SirenCustomClass::SetAttribute(const StringRef& val)
{
	return mAttribute.Load(val);
}

BaseSirenCustomType* SirenCustomClass::FindType(const StringRef& name) const
{
	if (name.Contains('.'))
	{
		List<StringRef> outPaths;
		StringParser::Split(name, ".", outPaths, false);
		if (outPaths.Count() <= 1)
		{
			return nullptr;
		}

		auto* rootType = mTypes.GetOptional(outPaths[0], nullptr);
		if (rootType == nullptr || !rootType->IsCustomClass())
		{
			return nullptr;
		}

		SirenCustomClass* parentType = (SirenCustomClass*)rootType;
		for (size_t i = 1; i < outPaths.Count(); ++i)
		{
			auto* childType = parentType->FindType(outPaths[i]);
			if (childType == nullptr || !childType->IsCustomClass())
			{
				return nullptr;
			}
			parentType = (SirenCustomClass*)childType;
		}

		return parentType;
	}
	else
	{
		return mTypes.GetOptional(name, nullptr);
	}
}

bool SirenCustomClass::HasType(const StringRef& name) const
{
	return mTypes.ContainsKey(name);
}

size_t SirenCustomClass::GetFieldCountRecursively() const
{
	if (mBaseType != nullptr)
	{
		return mBaseType->GetFieldCountRecursively() + mFields.Count();
	}
	return mFields.Count();
}

bool SirenCustomClass::Link(SirenAssembly& assembly)
{
	if (!mBaseTypeName.IsEmpty() && mBaseType == nullptr)
	{
		ISirenType* baseType = assembly.FindTypeWithReference(mBaseTypeName);
		if (baseType == nullptr)
		{
			Log::FormatError("Cannot find base class:{}", mBaseTypeName);
			return false;
		}
		if (!baseType->IsCustomClass())
		{
			Log::FormatError("{} cannot be a base class", mBaseTypeName);
			return false;
		}
		mBaseType = (SirenCustomClass*)baseType;
	}

	for (auto& typePair : mTypes)
	{
		auto* type = typePair.Value;
		RETURN_FALSE_IF_FALSE(type->Link(assembly));
	}

	for (auto& fieldPair : mFields)
	{
		auto* field = fieldPair.Value;
		RETURN_FALSE_IF_FALSE(field->Link(assembly));

		TryAddIncludeType(field->Type());
		TryAddIncludeType(field->KeyType());
		TryAddIncludeType(field->ValueType());
	}

	return true;
}

bool SirenCustomClass::IsCompleted() const
{
	RETURN_FALSE_IF(!mBaseTypeName.IsEmpty() && mBaseType == nullptr);

	for (auto& typePair : mTypes)
	{
		auto* type = typePair.Value;
		RETURN_FALSE_IF_FALSE(type->IsCompleted());
	}

	for (auto& fieldPair : mFields)
	{
		auto* field = fieldPair.Value;
		RETURN_FALSE_IF_FALSE(field->IsCompleted());
	}

	return true;
}

bool SirenCustomClass::Merge(const BaseSirenCustomType& other)
{
	if (!other.IsCustomClass())
	{
		return false;
	}

	SirenCustomClass& otherCustom = (SirenCustomClass&)other;
	for (auto& typePair : otherCustom.mTypes)
	{
		auto* curType = FindType(typePair.Key);
		if (curType == nullptr)
		{
			AddType((BaseSirenCustomType*)typePair.Value->Clone());
		}
		else
		{
			auto* otherType = typePair.Value;
			if (!curType->Merge(*otherType))
			{
				Log::FormatError("Cannot merge class:{}", typePair.Key);
				return false;
			}
		}
	}

	return true;
}

SirenCustomClass* SirenCustomClass::Clone() const
{
	SirenCustomClass* clone = new SirenCustomClass();
	clone->mName = mName;
	clone->mFullName = mFullName;
	clone->mBaseTypeName = mBaseTypeName;

	for (auto filePair:mFields)
	{
		clone->mFields.Add(filePair.Key, filePair.Value->Clone());
	}

	for (auto typePair : mTypes)
	{
		clone->mTypes.Add(typePair.Key, (BaseSirenCustomType*)typePair.Value->Clone());
	}

	return clone;
}

bool SirenCustomClass::LoadFrom(IStream& stream)
{
	RETURN_FALSE_IF_FALSE(BaseSirenCustomType::LoadFrom(stream));
	RETURN_FALSE_IF_FALSE(mAttribute.LoadFrom(stream));

	mBaseTypeName = stream.ReadString();

	//types
	uint typeCount = stream.Read<uint32>();
	FOR_EACH_SIZE(i, typeCount)
	{
		char isClass = (char)stream.ReadChar();
		if (isClass == 1)
		{
			std::unique_ptr<SirenCustomClass> type(new SirenCustomClass());
			RETURN_FALSE_IF_FALSE(type->LoadFrom(stream));
			type->SetParentType(this);
			mTypes.Add(type->Name(), type.get());
			type.release();
		}
		else
		{
			std::unique_ptr<SirenCustomEnum> type(new SirenCustomEnum());
			RETURN_FALSE_IF_FALSE(type->LoadFrom(stream));
			type->SetParentType(this);
			mTypes.Add(type->Name(), type.get());
			type.release();
		}
	}

	//fields
	uint fieldCount = stream.Read<uint32>();
	FOR_EACH_SIZE(i, fieldCount)
	{
		std::unique_ptr<SirenField> field(new SirenField());
		RETURN_FALSE_IF_FALSE(field->LoadFrom(stream));
		field->SetParentType(this);
		mFields.Add(field->Name(), field.get());
		field->SetIndex((ushort)mFields.Count()-1);
		field.release();
	}

	return true;
}


bool SirenCustomClass::SaveTo(IStream& stream) const
{
	RETURN_FALSE_IF_FALSE(BaseSirenCustomType::SaveTo(stream));
	RETURN_FALSE_IF_FALSE(mAttribute.SaveTo(stream));

	stream.WriteString(mBaseTypeName);

	//types
	uint typeCount = (uint)mTypes.Count();
	stream.Write(typeCount);
	for (auto& typePair : mTypes)
	{
		auto* type = typePair.Value;
		char isClass = type->IsCustomClass() ? 1 : 0;
		stream.WriteChar(isClass);
		type->SaveTo(stream);
	}

	//fields
	uint fieldCount = (uint)mFields.Count();
	stream.Write(fieldCount);
	for (auto& fieldPair : mFields)
	{
		auto* field = fieldPair.Value;
		field->SaveTo(stream);
	}

	return true;
}

bool SirenCustomClass::AddField(SirenField* val)
{
	if (mFields.TryAdd(val->Name(), val))
	{
		val->SetIndex((ushort)mFields.Count()-1);
		val->SetParentType(this);
		return true;
	}

	Log::FormatError("Duplicate property name:{}", val->Name());
	return false;
}

bool SirenCustomClass::AddType(BaseSirenCustomType* val)
{
	if (mTypes.TryAdd(val->Name(), val))
	{
		val->SetParentType(this);
		return true;
	}

	Log::FormatError("Duplicate child class name:{}", val->Name());
	return false;
}


bool SirenCustomClass::TryAddIncludeType(ISirenType* val)
{
	RETURN_FALSE_IF_NULL(val);
	if (val->IsBuildIn())
	{
		return false;
	}

	BaseSirenCustomType* customType = (BaseSirenCustomType*)val;
	mIncludeTypes.TryAdd(customType);

	return true;
}

bool SirenCustomClass::AddAttribute(StringRef name, StringRef val)
{
	return mAttribute.AddAttribute(name, val);
}

bool SirenCustomClass::Parse(SirenAssembly& assembly, StringRef& refProto)
{
	mName = SirenTextParser::ReadToken(refProto);
	if (mName.IsEmpty())
	{
		Log::Error("Cannot get class name");
		return false;
	}

	char c = SirenTextParser::ReadNextPrintChar(refProto);
	if (c == '\0')
	{
		Log::Error("Invalid class declare");
		return false;
	}

	if (c == ':')
	{
		//may has base class
		mBaseTypeName = SirenTextParser::ReadTypeName(refProto);
		if (mBaseTypeName.IsEmpty())
		{
			Log::Error("Cannot get base class.");
			return false;
		}


		c = SirenTextParser::ReadNextPrintChar(refProto);
		if (c == '\0')
		{
			Log::Error("Invalid class declare");
			return false;
		}
	}

	if (c == '[')
	{
		StringRef attributeStr = SirenTextParser::ReadAttribute(refProto);
		if (attributeStr.IsEmpty())
		{
			Log::Error("Cannot get attribute.");
			return false;
		}

		if (!SetAttribute(attributeStr))
		{
			Log::FormatError("Invalid class attribute:{}", attributeStr);
			return false;
		}

		c = SirenTextParser::ReadNextPrintChar(refProto);
		if (c == '\0')
		{
			Log::Error("Invalid class declare");
			return false;
		}

	}

	if (c == '{')
	{
		while (true)
		{
			StringRef token = SirenTextParser::ReadTypeName(refProto);
			if (token.IsEmpty())
			{
				break;
			}

			if (token == SirenTextParser::EnumKeyword)
			{
				std::unique_ptr<SirenCustomEnum> child(new SirenCustomEnum());
				if (child->Parse(assembly, refProto))
				{
					RETURN_FALSE_IF_FALSE(AddType(child.get()));
					child.release();
				}
				else
				{
					Log::Error("Failed to parse enum.");
					return false;
				}
			}
			else if (token == SirenTextParser::ClassKeyword)
			{
				std::unique_ptr<SirenCustomClass> child(new SirenCustomClass());
				if (child->Parse(assembly, refProto))
				{
					RETURN_FALSE_IF_FALSE(AddType(child.get()));
					child.release();
				}
				else
				{
					Log::Error("Failed to parse class.");
					return false;
				}
			}
			else
			{
				//properties
				std::unique_ptr<SirenField> field(new SirenField());
				if (field->Parse(assembly, this, token, refProto))
				{
					RETURN_FALSE_IF_FALSE(AddField(field.get()));
					field.release();
				}
				else
				{
					Log::Error("Failed to parse property.");
					return false;
				}
			}
		}
	}
	else
	{
		Log::Error("Invalid class declare");
		return false;
	}



	return SirenTextParser::EndType(refProto);

}

MEDUSA_END;

File: Extension/Medusa/Medusa/Siren/Schema/SirenCustomClass.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using Medusa.Common;

namespace Medusa.Siren.Schema
{
    public class SirenCustomClass : BaseSirenCustomType
    {
        public override bool IsCustomClass => true;
        public override bool IsCustomEnum => false;

        public override SirenTypeId Id => SirenTypeId.Struct;
        public SirenClassAttribute Attribute { get; set; }
        public List<SirenField> Fields { get; private set; }
        public Dictionary<string, SirenField> FieldNameDict { get; private set; }
        public Dictionary<ushort, SirenField> FieldIdDict { get; private set; }

        public Dictionary<string, BaseSirenCustomType> Types { get; set; }

        public string BaseTypeName { get; set; }
        public SirenCustomClass BaseType { get; set; }

        public MethodInfo SerializeMethodInfo { get; private set; }
        public MethodInfo DeserializeMethodInfo { get; private set; }
        public MethodInfo RegisterMethodInfo { get; private set; }

        public SirenCustomClass(Type type, SirenClassAttribute attribute)
        {
            Type = type;
            Attribute = attribute;
            Fields = new List<SirenField>();
            FieldNameDict = new Dictionary<string, SirenField>();
            FieldIdDict = new Dictionary<ushort, SirenField>();
        }

        public SirenCustomClass(string name = "") : base(name)
        {
            Fields = new List<SirenField>();
            FieldNameDict = new Dictionary<string, SirenField>();
            FieldIdDict = new Dictionary<ushort, SirenField>();
        }


        public override object Clone()
        {
            SirenCustomClass val = new SirenCustomClass(Name) { Attribute = Attribute.Clone() as SirenClassAttribute };
            val.BaseTypeName = BaseTypeName;
            val.BaseType = BaseType.Clone() as SirenCustomClass;
            foreach (var baseSirenCustomType in Types)
            {
                val.Types.Add(baseSirenCustomType.Key, baseSirenCustomType.Value.Clone() as BaseSirenCustomType);
            }

            foreach (var sirenField in FieldNameDict)
            {
                val.FieldNameDict.Add(sirenField.Key, sirenField.Value.Clone() as SirenField);
            }
          

            return val;
        }

        public override void AddAttribute(string name, string val)
        {
            Attribute.AddAttribute(name, val);
        }

        public override void SetAttribute(string val)
        {
            Attribute.Load(val);
        }

        public override bool LoadFrom(Stream stream)
        {
            base.LoadFrom(stream);
            Attribute.LoadFrom(stream);
            BaseTypeName = stream.ReadString();

            //types
            uint typeCount = stream.ReadUInt();
            for (int i = 0; i < typeCount; i++)
            {
                byte isClass = (byte)stream.ReadByte();
                if (isClass == 1)
                {
                    SirenCustomClass type = new SirenCustomClass();
                    type.LoadFrom(stream);
                    type.Parent = this;
                    Types.Add(type.Name, type);
                }
                else
                {
                    SirenCustomEnum type = new SirenCustomEnum();
                    type.LoadFrom(stream);
                    type.Parent = this;
                    Types.Add(type.Name, type);
                }
            }
            //fields
            uint fieldCount = stream.ReadUInt();
            for (int i = 0; i < fieldCount; i++)
            {
                SirenField field = new SirenField();
                field.LoadFrom(stream);
                field.ParentType = this;
                FieldNameDict.Add(field.Name, field);
                field.Index = (ushort)(FieldNameDict.Count - 1);
            }

            return true;
        }

        public override bool SaveTo(Stream stream)
        {
            base.SaveTo(stream);
            Attribute.SaveTo(stream);
            stream.WriteString(BaseTypeName);

            //types
            uint typeCount = (uint)Types.Count;
            stream.Write(typeCount);
            foreach (var baseSirenCustomType in Types)
            {
                byte isClass = baseSirenCustomType.Value.IsCustomClass ? (byte)1 : (byte)0;
                stream.WriteByte(isClass);
                baseSirenCustomType.Value.SaveTo(stream);
            }

            //fields
            uint filedCount = (uint)FieldNameDict.Count;
            stream.Write(filedCount);
            foreach (var sirenField in FieldNameDict)
            {
                sirenField.Value.SaveTo(stream);
            }


            return true;
        }

        public override bool Link(SirenAssembly assembly)
        {
            return false;
        }

        public bool Initialize()
        {
            if (Type.IsEnum)
            {
                return true;
            }

            //get methods
            SerializeMethodInfo = Type.GetMethod("Serialize", BindingFlags.Public | BindingFlags.Instance);
            DeserializeMethodInfo = Type.GetMethod("Deserialize", BindingFlags.Public | BindingFlags.Instance);
            RegisterMethodInfo = Type.GetMethod("Register", BindingFlags.Public | BindingFlags.Instance | BindingFlags.Static);
            if (RegisterMethodInfo != null && !Type.ContainsGenericParameters)
            {
                RegisterMethodInfo.Invoke(null, null);
            }


            //get properties
            uint index = 0;
            ushort id = GetBasePropertyCount();
            var properties = Type.GetProperties(BindingFlags.DeclaredOnly | BindingFlags.Public | BindingFlags.Instance);
            foreach (var propertyInfo in properties)
            {
                var attrs = propertyInfo.GetCustomAttributes(typeof(SirenFieldAttribute), false);
                if (attrs.Length > 0)
                {
                    SirenFieldAttribute propertyAttribute = attrs[0] as SirenFieldAttribute;
                    if (propertyAttribute != null)
                    {
                        SirenField property = new SirenField(this, propertyInfo, propertyAttribute, index++, id++);
                        Fields.Add(property);
                        FieldIdDict.Add(property.Id, property);
                        FieldNameDict.Add(property.Name, property);
                    }
                }

            }

            if (Fields.Count == 0)
            {
                if (Type.BaseType == typeof(object) || Type.BaseType.IsValueType)
                {
                    return false;
                }
            }

            if (!Type.IsValueType&& Type.BaseType != typeof(object) && !Type.BaseType.IsValueType)
            {
                BaseType = SirenMachine.GetClass(Type.BaseType);
                if (BaseType != null)
                {
                    //add base properties
                    foreach (var sirenProperty in BaseType.FieldIdDict)
                    {
                        FieldIdDict.Add(sirenProperty.Key, sirenProperty.Value);
                    }

                    foreach (var sirenProperty in BaseType.FieldNameDict)
                    {
                        FieldNameDict.Add(sirenProperty.Key, sirenProperty.Value);
                    }
                }
            }


            return true;
        }

        public ushort GetBasePropertyCount()
        {
            if (Type.BaseType != typeof(object))
            {
                var baseClass = SirenMachine.FindClass(Type.BaseType);
                if (baseClass != null)
                {
                    return (ushort)(baseClass.GetBasePropertyCount() + (ushort)baseClass.Fields.Count);
                }
            }
            return 0;
        }

        public SirenField FindField(ushort id)
        {
            SirenField outProperty;
            FieldIdDict.TryGetValue(id, out outProperty);
            return outProperty;
        }

        public SirenField FindField(String name)
        {
            SirenField outProperty;
            FieldNameDict.TryGetValue(name, out outProperty);
            return outProperty;
        }
    }
}
